### ISSUE TITLE
[Flight] Define GRPC IDL / wire protocol for messaging with Arrow data

### ISSUE THREAD

**Issue by toddfarmer** [toddfarmer/arrow-migration#2218]
***Note**: This issue was originally created as [ARROW-249](https://issues.apache.org/jira/browse/ARROW-249). Please see the [migration documentation](https://gist.github.com/toddfarmer/12aa88361532d21902818a6044fda4c3) for further details.*

### Original Issue Description:
In addition to memory maps / shared memory, we should be able to assemble one or more Arrow arrays in a memory block suitable for sending over the wire with GRPC (http://www.grpc.io/). This can be similarly adapted to other messaging / RPC frameworks. 

We can continue to use Flatbuffers for the metadata serialization. 
### Related issues:
 - #2228 (requires)

### Migrated issue participants:
**Reporter**: Wes McKinney (wesm)
**Assignee**: Jacques Nadeau (jnadeau)

**Comment by toddfarmer** [comment]
***Note**: [Comment](https://issues.apache.org/jira/browse/ARROW-249?focusedCommentId=15418020) by Julien Le Dem (julienledem):*
Some requirements for the RPC layer:
 - When receiving a message we need to be able to address the buffer where the message is directly from the RPC layer without copying it.
 - when sending a message, we need to be able to point to several Arrow buffers and have them sent directly (one after the other) with no copy as well.

**Comment by toddfarmer** [comment]
***Note**: [Comment](https://issues.apache.org/jira/browse/ARROW-249?focusedCommentId=15418045) by Wes McKinney (wesm):*
Agreed. I do not know enough about GRPC to know how much protobuf overhead there is on receipt. If this returns a buffer of bytes in the deserialized protobuf then this can be accessed via Flatbuffers with zero copy.

**Comment by toddfarmer** [comment]
***Note**: [Comment](https://issues.apache.org/jira/browse/ARROW-249?focusedCommentId=15418072) by Jacques Nadeau (jnadeau):*
I remember some from Kudu team (maybe `[~tlipcon]` or `[~jdcryans]`?) mentioned something about how they had to customize the GRPC for a sidecar data batch and see some mention of it here: https://github.com/grpc/grpc-java/issues/1054. I wonder if they could share any insights.

We implemented something custom in Drill on top of Netty before GRPC was built to satisfy the need to support off-heap sideband messages but it would be nice to have someone else manage the maintenance of the layer.

**Comment by toddfarmer** [comment]
***Note**: [Comment](https://issues.apache.org/jira/browse/ARROW-249?focusedCommentId=16009556) by Wes McKinney (wesm):*
I'd like to prototype a GRPC-based client and server that sends components of an Arrow stream as a series of RPC messages according to the Arrow streaming format. See this discussion about some issues around this

https://github.com/grpc/grpc-java/issues/1054#issuecomment-301282237

In ARROW-1008, I've refactored the "stream reader" and "stream writer" concepts to be abstract interfaces, so now we could define `GrpcStreamWriter` and `GrpcStreamReader` that would be a middleman to the underlying GRPC client/server machinery, presenting a standard stream iterator to the client side. I figure we could prototype this using protobuf messages that hold each piece of the Arrow stream (first RPC message is the schema) and see how the perf compares with a "raw" stream. 

cc `[~nongli]`

**Comment by toddfarmer** [comment]
***Note**: [Comment](https://issues.apache.org/jira/browse/ARROW-249?focusedCommentId=16197330) by Jacques Nadeau (jnadeau):*
Some key performance requirements from my perspective (note that I'm just skipping all the basic requirements):

- Send N sidecar ByteBufs in addition to a structured message on the sending side, only doing a single copy: the kernel copying the ByteBuf to the socket.
- Receive and pull out a single sidecar ByteBuf on the receiving-side, keeping it off heap and minimizing for its lifetime. Preferably zero copies but potentially a single copy (from a circular or rotating buffer to a final target buffer).
- Support a cooperative backpressure methodology when using multiple streams on a single connection.
- Handle out of direct memory situations without disrupting the connection (message failure as opposed to connection failure).
  
  I'm trying to piece together how we might do this using some less than ideal usage of the GRPC apis. Not sure how bad the shenaniganry will need to be. Note that we have something we've built up at Dremio (and before) that we've found very reliable for the performance requirements above:
  
- https://github.com/dremio/dremio-oss/tree/master/services/base-rpc
  
  Obviously, this is less well supported/contributed/featured than GRPC but used at many large scale production customers already. It doesn't do all the things that GRPC does but it does do the requirements above very well. On the flipside, it doesn't have all the client/language support, doesn't have all the security features, compression features, etc.
